pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/TypeHelper.java

```diff
@@ -24,28 +24,55 @@ private TypeHelper() {
      * @return <code>true</code> if type node n is of type clazzName or a subtype of clazzName
      */
     public static boolean isA(final TypeNode n, final String clazzName) {
+        final Class<?> clazz = loadClassWithNodeClassloader(n, clazzName);
+
+        if (clazz != null) {
+            return isA(n, clazz);
+        }
+
+        return clazzName.equals(n.getImage()) || clazzName.endsWith("." + n.getImage());
+    }
+    
+    /**
+     * Checks whether the resolved type of the given {@link TypeNode} n is exactly of the type
+     * given by the clazzName.
+     *
+     * @param n the type node to check
+     * @param clazzName the class name to compare to
+     * @return <code>true</code> if type node n is exactly of type clazzName.
+     */
+    public static boolean isExactlyA(final TypeNode n, final String clazzName) {
+        final Class<?> clazz = loadClassWithNodeClassloader(n, clazzName);
+
+        if (clazz != null) {
+            return n.getType() == clazz;
+        }
+
+        return clazzName.equals(n.getImage()) || clazzName.endsWith("." + n.getImage());
+    }
+    
+    private static Class<?> loadClassWithNodeClassloader(final TypeNode n, final String clazzName) {
         if (n.getType() != null) {
             try {
                 ClassLoader classLoader = n.getType().getClassLoader();
                 if (classLoader == null) {
                     // Using the system classloader then
                     classLoader = ClassLoader.getSystemClassLoader();
                 }
-
+    
                 // If the requested type is in the classpath, using the same classloader should work
-                final Class<?> clazz = classLoader.loadClass(clazzName);
-
-                if (clazz != null) {
-                    return isA(n, clazz);
-                }
+                return classLoader.loadClass(clazzName);
             } catch (final ClassNotFoundException ignored) {
                 // The requested type is not on the auxclasspath. This might happen, if the type node
                 // is probed for a specific type (e.g. is is a JUnit5 Test Annotation class).
                 // Failing to resolve clazzName does not necessarily indicate an incomplete auxclasspath.
```

Comment:
It seems to me like it contradicts the comment two lines above. I don't see how this couldn't indicate an incomplete classpath, and hence why we shouldn't report it.

Reply:
The class being loaded is the class you want to test against (ie: is this Guava's Optional?) but if that class is not present in the current auxclasspath it may well be that the code doesn't use Guava at all, and this type will not be Guava's Optional. There is nothing wrong with the project itself (at most, you are enabling a rule that makes no sense in this particular project).

I'd not warn about that either non-the-less, companies sometimes have a single standard ruleset to apply to all projects regardless of specifics (ie: if you use Guava I want to check for this...)

Reply:
Aah I see. Thanks for the explanation. You're right it makes sense to not report